io-sim/src/Control/Monad/IOSimPOR/Internal.hs

```diff
@@ -0,0 +1,1574 @@
+    -- from `GHC.Event` behaviour.
+    NewTimeout d k | d < 0 -> do
+      let t       = NegativeTimeout nextTmid
+          expiry  = d `addTime` time
+          thread' = thread { threadControl = ThreadControl (k t) ctl }
+      trace <- schedule thread' simstate { nextTmid = succ nextTmid }
+      return (SimTrace time tid tlbl (EventTimerCreated nextTmid nextVid expiry) $
+              SimTrace time tid tlbl (EventTimerCancelled nextTmid) $
+              trace)
+
+    NewTimeout d k -> do
+      tvar  <- execNewTVar nextVid
+                           (Just $ "<<timeout-state " ++ show (unTimeoutId nextTmid) ++ ">>")
+                           TimeoutPending
+      tvar' <- execNewTVar (succ nextVid)
+                           (Just $ "<<timeout " ++ show (unTimeoutId nextTmid) ++ ">>")
+                           False
+      let expiry  = d `addTime` time
+          t       = Timeout tvar tvar' nextTmid
+          timers' = PSQ.insert nextTmid expiry (TimerVars tvar tvar') timers
+          thread' = thread { threadControl = ThreadControl (k t) ctl }
+      trace <- schedule thread' simstate { timers   = timers'
+                                         , nextVid  = succ (succ nextVid)
+                                         , nextTmid = succ nextTmid }
+      return (SimTrace time tid tlbl (EventTimerCreated nextTmid nextVid expiry) trace)
+
+    -- we do not follow `GHC.Event` behaviour here; updating a timer to the past
+    -- effectively cancels it.
+    UpdateTimeout (Timeout _tvar _tvar' tmid) d k | d < 0 -> do
+      let timers' = PSQ.delete tmid timers
+          thread' = thread { threadControl = ThreadControl k ctl }
+      trace <- schedule thread' simstate { timers = timers' }
+      return (SimTrace time tid tlbl (EventTimerCancelled tmid) trace)
+
+    UpdateTimeout (Timeout _tvar _tvar' tmid) d k -> do
+          -- updating an expired timeout is a noop, so it is safe
+          -- to race using a timeout with updating or cancelling it
+      let updateTimeout_  Nothing       = ((), Nothing)
+          updateTimeout_ (Just (_p, v)) = ((), Just (expiry, v))
+          expiry  = d `addTime` time
+          timers' = snd (PSQ.alter updateTimeout_ tmid timers)
+          thread' = thread { threadControl = ThreadControl k ctl }
+      trace <- schedule thread' simstate { timers = timers' }
+      return (SimTrace time tid tlbl (EventTimerUpdated tmid expiry) trace)
+
+    -- updating a negative timer is a no-op, unlike in `GHC.Event`.
+    UpdateTimeout (NegativeTimeout _tmid) _d k -> do
+      let thread' = thread { threadControl = ThreadControl k ctl }
+      schedule thread' simstate
+
+    CancelTimeout (Timeout _tvar _tvar' tmid) k -> do
+      let timers' = PSQ.delete tmid timers
+          thread' = thread { threadControl = ThreadControl k ctl }
+      trace <- schedule thread' simstate { timers = timers' }
+      return (SimTrace time tid tlbl (EventTimerCancelled tmid) trace)
```

Comment:
Interesting. I don't think I changed this, and I don't see anything IOSimPOR specific in there. I did work from a copy of IOSim, perhaps this changed after I took that copy? Looking at the code that is in IOSim now, it's waking up threads and would certainly need to interact with the race detection, so it can't simply be copied into IOSimPOR as it is.